clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go

```diff
@@ -23,178 +23,419 @@ package common
+			if !ok {
 				return
 			}
-
-			atomic.AddInt32(&runningRuntimeObjects, 1)
-			objName, err := getNameFromRuntimeObject(runtimeObjectsList[objectIndex])
+			resourceVersion, err := getResourceVersionFromRuntimeObject(runtimeObj)
 			if err != nil {
-				errList.Append(fmt.Errorf("reading object name error: %v", err))
 				return
 			}
-			objNamespace, err := getNamespaceFromRuntimeObject(runtimeObjectsList[objectIndex])
+			w.updateOpResourceVersion(resourceVersion)
+			w.handleObject(c, runtimeObj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldRuntimeObj, ok := oldObj.(runtime.Object)
+			if !ok {
+				return
+			}
+			newRuntimeObj, ok := newObj.(runtime.Object)
+			if !ok {
+				return
+			}
+			resourceVersion, err := getResourceVersionFromRuntimeObject(newRuntimeObj)
 			if err != nil {
-				errList.Append(fmt.Errorf("reading object namespace error: %v", err))
 				return
 			}
-			glog.Infof("%s: %s (%s) has all pods running", waitForControlledPodsRunningName, objName, objNamespace)
-		})
+			// Increment version regardless of change in spec.
+			w.updateOpResourceVersion(resourceVersion)
+			oldSpec, err := getSpecFromRuntimeObject(oldRuntimeObj)
+			if err != nil {
+				return
+			}
+			newSpec, err := getSpecFromRuntimeObject(newRuntimeObj)
+			if err != nil {
+				return
+			}
+			if reflect.DeepEqual(oldSpec, newSpec) {
+				// If spec of the object hasn't change, do nothing.
+				return
+			}
+			w.handleObject(c, newRuntimeObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			runtimeObj, ok := obj.(runtime.Object)
+			if !ok {
+				return
+			}
+			resourceVersion, err := getResourceVersionFromRuntimeObject(runtimeObj)
+			if err != nil {
+				return
+			}
+			w.updateOpResourceVersion(resourceVersion)
+			w.deleteObject(runtimeObj)
+		},
+	})
+
+	go w.informer.Run(w.stopCh)
+	timeoutCh := make(chan struct{})
+	timeoutTimer := time.AfterFunc(informerSyncTimeout, func() {
+		close(timeoutCh)
+	})
+	if !cache.WaitForCacheSync(timeoutCh, w.informer.HasSynced) {
+		return fmt.Errorf("timed out waiting for caches to sync")
 	}
-	wg.Wait()
+	timeoutTimer.Stop()
 
-	if int(runningRuntimeObjects) == len(runtimeObjectsList) {
-		glog.Infof("%s: %d / %d %ss are running with all pods", waitForControlledPodsRunningName, runningRuntimeObjects, len(runtimeObjectsList), kind)
-		return summaries, nil
+	store := w.informer.GetStore()
+	objects := store.List()
+	for i := range objects {
+		runtimeObj, ok := objects[i].(runtime.Object)
+		if !ok {
+			continue
+		}
+		w.handleObject(c, runtimeObj)
+	}
+	return nil
+}
+
+func (w *waitForControlledPodsRunningMeasurement) gather(timeout time.Duration) error {
+	glog.Infof("%v: waiting for controlled pods measurement...", waitForControlledPodsRunningName)
+	if w.informer == nil {
+		return fmt.Errorf("metric %s has not been started", waitForControlledPodsRunningName)
+	}
+	startTime := time.Now()
+	store := w.informer.GetStore()
+	objects := store.List()
```

Comment:
This is incorrect. The list should be done from apiserver directly - the cache may be lagging.

Reply:
Done.